Update sum types design (#2187)
carbon-language/carbon-lang: docs/design/sum_types.md

# Sum types

<!--
Part of the Carbon Language project, under the Apache License v2.0 with LLVM
Exceptions. See /LICENSE for license information.
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

<!-- toc -->

## Table of contents

-   [Overview](#overview)
-   [`choice` declarations](#choice-declarations)
-   [User-defined sum types](#user-defined-sum-types)
-   [Alternatives considered](#alternatives-considered)
-   [References](#references)

<!-- tocstop -->

> ## Overview

Reviewer: Maybe it'd be helpful to draw a comparison to `enum`s? This document currently only covers the more complex cases of `choice`, not the `enum`-equivalent  scenario.

Author: I'm ambivalent about that. It could be a useful illustration, but it also opens up a can of worms: `choice` isn't a complete substitute for C++ `enum`, because a substantial fraction of `enum` use cases rely on the ability to convert to/from integer types, and/or the fact that an `enum` can have a value other than one of the enumerators. So it could be helpful from a teaching point of view, but it could also be counterproductive from a specification point of view.

In Carbon, a _sum type_ is a type whose values are grouped into several distinct

> named cases, called _alternatives_. A value of a sum type logically consists of

Reviewer: What do you mean with "logically" here? Are there other words that could make the intent clearer?

Author: I mean "logically" as opposed to "physically". For example, a nullable pointer is a valid implementation of `Optional(T*)` even though the discriminator doesn't physically exist in memory -- the discriminator logically exists, because you can compute its value from the representation of the nullable pointer.

I don't want to spend a paragraph trying to explain all that, though, because it would be a distraction; I just want to try to head off any misunderstanding that this means the discriminator and the alternative are physically separate values in memory. I suppose I could say something like "conceptually", "notionally", or "in principle" instead of "logically", but I'm not sure any of those is an improvement.

Reviewer: I'd suggest "conceptually" or "notionally". "logically" could also mean "characterized by or capable of clear, sound reasoning" and so using something less ambiguous in meaning would be helpful.

Author: OK, switched to "notionally".

a value of one of the alternatives, together with a _discriminator_ tag, which

> identifies which alternative is present. Sum types are typically handled with

Reviewer: I think there are extra commas here, making it hard to read: is "together with a discriminator tag" intended to be a sub-clause, with "which identifies which alternative is present" referring to the value of one of the alternatives? Or is the third part referring to the discriminator tag?

Consider rephrasing, e.g.:

...a value of one of the alternatives together with a _discriminator_ tag. The discriminator...

However, as I think about this further, is this a good explanation? Consider:

```
choice Foo {
  Bar,
  Baz,
}
```

This is only the discriminator, no additional values, right? Perhaps it would be better to say:

A value of a sum type consists of a _discriminator_ tag which identifies which alternative is present, as well as the alternative's value if it has one.

Author: Yeah, that seems better. Done.

(I might argue that formally the alternative always has a value, although sometimes it's a value of an empty type like `()`, but splitting that hair isn't helpful here)

pattern matching.

## `choice` declarations

The `choice` keyword is used to declare a sum type by specifying its interface,
leaving the implementation to the compiler. For example:

```carbon
choice OptionalI32 {
  Some(value: i32),
  None
}
```

This declares a sum type named `OptionalI32` with two alternatives: `Some`,
which holds a single `i32` value, and `None`, which is empty. Choice types can

> also be parameterized, like class types:

Reviewer: Can you link to the class type explanation that you're comparing to?

Author: Done

```carbon
choice Optional(T:! Type) {
  Some(value: T),
  None
}
```

> An alternative declaration consists of the alternative name, followed by an

Reviewer: Consider moving this explanation up. In other words, rather than giving examples first, explain the syntax first.

Author: Done.

> optional parameter list in parentheses, much like the syntax of a function

Reviewer: Can you link to the syntax you're referring to?

Author: Done.

> declaration. A value of a function-like alternative is specified by "calling" it

Reviewer: The comma before "followed" makes it look like a sub-clause, implying that the alternative name is like the syntax of a function declaration. I don't think that's correct.

On the other reading, it says that the optional list of parameters is similar to the syntax of a function declaration. I don't think that's correct either: parameter lists are never optional for function declarations.

Maybe say something like:

An alternative declaration always has a name, and may have a parameter list. When present, the parameter list uses function-like parameter syntax.

Author: How's this?

> like a function, and a value of a "bare" alternative like `None` is specified by

Reviewer: This introduces the concept of "bare" whereas line 42 says "empty". Would it make sense to use a single term through the doc?

Author: Yeah, and p0157 uses "bare" to mean something else, so I've standardized on "empty".

be extended to behave like a sum type. This is much more verbose than a `choice`
declaration, but gives the author full control over the representation and class
members.

The ability to create instances of the sum type can be straightforwardly
emulated with factory functions and static constants (although the latter
[are not yet designed](classes.md#no-static-variables)), and the internal
storage layout will presumably involve untagged unions or some other low-level
storage primitive which hasn't been designed yet, but the key to defining a sum
type's interface is enabling it to support pattern matching. To do that, the sum
type has to specify two things:

-   The set of all possible alternatives, including their names and parameter
    types, so that the compiler can typecheck the `match` body, identify any
    unreachable `case`s, and determine whether any `case`s are missing.
-   The algorithm that, given a value of the sum type, determines which
    alternative is present, and specifies the values of its parameters.

Here's how that would look if `Optional` were defined as a class:

```carbon
class Optional(T:! Type) {
  // Factory functions
  fn Some(value: T) -> Self;

>   fn None() -> Self;

Reviewer: ```suggestion
  let None: Self;
```

I think the function syntax here is inconsistent with the `choice` equivalent. Or is the skew deliberate? If so, maybe it should be clarified that `Optional` as a class is going to inevitably differ?

Author: Yeah, this is better, I just hadn't realized such a declaration was permitted. Done.

  private var has_value: bool;
  private var value: T;

>   interface MatchContinuation {

Reviewer: Is `MatchContinuation` specifically designed somewhere?

Author: The design of `Optional.MatchContinuation` is part of the design of `Optional`. For present purposes, `Optional` is just an example, so it doesn't need a design.

Keep in mind, the role of this interface is sort of "backwards" compared to the interfaces that define things like overloadable operators. Typically we allow libraries to extend the core language by defining an interface that the library implements. Here, the situation is inverted: the library is defining an interface, which the compiler then implements. So, in designing the language, we don't design the interface itself, we design how the compiler generates an implementation of that interface, which is discussed below.

>     let template ReturnType:! Type;

Reviewer: I'm confused here, would this work as: `let ReturnType: Optional(T)`?

Author: No, that wouldn't work. `ReturnType` names the return type of the continuation methods below, so its values must be types, and hence its type must be a type-of-types.

    fn Some(value: T) -> ReturnType;
    fn None() -> ReturnType;
  }

>   external impl as Match(MatchContinuation) {

Reviewer: The proposal used `Matchable`, not `Match`. Elsewhere we do things like `EqWith` so I don't think we should do this rename without a proposal.

Reviewer 2: The PR descriptions points to this change coming from #1058

Reviewer: `match` or `Matchable` is not mentioned on #1058.

AFAICT pointing to #1058 is a claim that Match is "Interfaces that contribute usefully to creating consistent APIs on types", leading to Matchable. I would argue it's instead "Interfaces that implement a specific language expression syntax" (i.e., `match`), leading to `Match` (the accepted proposal state, and similar to `ImplicitAs`).

But to be clear, I think changes to things accepted in proposals should be done by proposals.

Reviewer: Sorry, got that a little backwards: `Matchable` is the accepted proposal state, and would be the consistent API interface. `Match` is one argument to name under the second rule.

But I'm not trying to argue about how #1058 could be read here -- we have an accepted proposal.

Author: OK, I've promoted this to a proposal.

For clarity: is your concern purely procedural, or are you arguing that `Match.Op` is not the right name under #1058?

Reviewer 3: My view is that `Match.Op` is a suitable name, consistent with the direction in #1058. I agree with @jonmeow that we should have this decision made in a proposal, so thanks for making this PR into one :)

    fn Op[me: Self, Continuation:! MatchContinuation](

>         continuation: Continuation*) {

Reviewer: Shouldn't this provide a return type?

Author: Whoops, yes. Good catch.

>         continuation: Continuation*) {

Reviewer: Is the `*` here needed? (I don't think continuation is being mutated)

If needed, should it be `addr continuation` instead?

Author: Whoops, this reminds me we need `Self` parameters on the methods of `MatchContinuation`, so I've done that.

> Is the `*` here needed? (I don't think continuation is being mutated)

`Continuation` is generated by the compiler to implement the `match` body, and it seems possible that the compiler might want it to be stateful, and even mutably stateful. Of course if we omit the `*`, the compiler can always make `Continuation` store a pointer to mutable state outside the class, but I'm reluctant to depart from the design of the accepted proposal without a clear rationale.

> If needed, should it be `addr continuation` instead?

I thought `addr` wasn't permitted on parameters other than `me`?

Reviewer: > I thought addr wasn't permitted on parameters other than me?

My mistake

      }
    }
  }

  // Operations like destruction, copying, assignment, and comparison are
  // omitted for brevity.
}
```

In this code, `Optional` makes itself available for use in pattern matching by
declaring that it implements the `Match` interface. `Match` takes an interface
argument, called the _continuation interface_, which specifies the set of
possible alternatives by declaring a method for each one. In this case, we pass
`Optional.MatchContinuation` as the continuation interface.

When compiling a `match` statement, the compiler checks that the type being
matched implements `Match(C)` for some continuation interface `C`. Then, it
notionally transforms the `match` body into a class that implements `C`, with
one method for each `case`, and then passes that to `Match.Op` on the object
being matched. For example, the `match` statement shown earlier might be
transformed into:

```carbon
class __MatchStatementImpl {
  fn Make() -> Self { return {}; }

  impl as Match(Optional.MatchContinuation) where .ReturnType = () {
    fn Some(the_value: i32) {
      Print(the_value);
    }
    fn None() {
      Print("None");
    }
  }
}

my_opt.(Match.Op)(__MatchStatementImpl.Make());

> ```

Reviewer: This seems like significant design since it seems to be injecting a name into the calling namespace, but I'm not finding it in the proposal. Can you please help me find where the __MatchStatementImpl design was discussed?

Author: This isn't intended to be a design change. See [here](https://github.com/carbon-language/carbon-lang/blob/57cc4c6dec942580388a164f226933ae586c4b90/proposals/p0157.md?plain=1#L445):

> ... an instance of the continuation interface, which the compiler generates from the `match` expression being evaluated, with method bodies that correspond to the bodies of the corresponding `case`s

`__MatchStatementImpl` is an illustration of what that instance might look like, if written in Carbon code.

Reviewer: Would it make sense to remove the `__`? I'm particularly wary of `__` because of its meaning in C++ and the discussion around #2029. For me, it implies that this is a language reserved name, and probably generated as part of the language, versus just sample code.

Author: This is sample _compiler generated_ code, so those connotations seem exactly right to me.

Reviewer: I'm pretty strongly opposed to Carbon generating code in user-owned namespaces using `__`. I believe the use of `__` in C++ reflects a design around generating code without using strong namespacing; in Carbon, I believe we can either (a) generate anonymous functions or (b) use namespaces owned by Carbon. In either case, injecting `__` into the name in order to prevent collisions seems unnecessary to me.

If you insist on using `__`, I'm going to suggest that this should be a question to leads: should the language be allowed to generate code inside of user-owned namespaces relying on `__` prefixes to disambiguate, or do we think other solutions are feasible+preferable?

Reviewer: (note, I believe that would be a corollary to #2029 -- but if the resolution of #2029 is that there `__` isn't reserved, then it seems like we also wouldn't use it for generating code)

Author: > I'm pretty strongly opposed to Carbon generating code in user-owned namespaces using `__`.

That's not my intent; this is just an illustration, since I can't directly depict the compiler-internal representation of an anonymous class. I've revised the text to try to make that clearer.

-->

<!-- toc -->

## Table of contents

-   [Overview](#overview)
-   [`choice` declarations](#choice-declarations)
-   [User-defined sum types](#user-defined-sum-types)
-   [Alternatives considered](#alternatives-considered)
-   [References](#references)

<!-- tocstop -->

## Overview

In Carbon, a _sum type_ is a type whose values are grouped into several distinct
named cases, called _alternatives_. A value of a sum type logically consists of
a _discriminator_ tag that identifies which alternative is present, together
with that alternative's value if it has one. Sum types are typically handled
with pattern matching.

## `choice` declarations

The `choice` keyword is used to declare a sum type by specifying its interface,
leaving the implementation to the compiler. A `choice` declaration consists of
the `choice` keyword followed by the name of the type, and then a list of
alternatives inside curly braces. An alternative declaration consists of the
alternative name, optionally followed by a parameter list in parentheses. If
present, the parameter list has the same syntax as in a
[function declaration](README.md#functions). For example:

```carbon
choice OptionalI32 {

>   Some(value: i32),

Reviewer 3: Are the parameter names used solely for documentation purposes? If so, I'd find it useful to say that somewhere near here. (And if not, I'd be interested in what else they do.)

Author: Done

  None
}
```

This declares a sum type named `OptionalI32` with two alternatives: `Some`,
which holds a single `i32` value, and `None`, which is empty. Choice types can
also be parameterized,
[like class types](generics/details.md#parameterized-types):

```carbon
choice Optional(T:! Type) {
  Some(value: T),
  None
}
```

A value of a function-like alternative is specified by "calling" it like a
function, and a value of an empty alternative like `None` is specified by naming
it:

```carbon
var my_opt: Optional(i32) = Optional(i32).None;
my_opt = Optional(i32).Some(42);
```

The value of a choice type can be inspected using a `match` statement:

```carbon
match (my_opt) {
  case .Some(the_value: i32) => {
    Print(the_value);
  }
  case .None => {
    Print("None");
  }
}
```

## User-defined sum types

`choice` declarations are a convenience shorthand for common use cases, but they
have limited flexibility. There is no way to control the representation of a
`choice` type, or define methods or other members for it (although you can

> extend it to implement interfaces, using

Reviewer 3: ```suggestion
extend it to implement interfaces, using an
```

declaration, but gives the author full control over the representation and class
members.

The ability to create instances of the sum type can be straightforwardly
emulated with factory functions and static constants, and the internal storage
layout will presumably involve untagged unions or some other low-level storage
primitive which hasn't been designed yet, but the key to defining a sum type's
interface is enabling it to support pattern matching. To do that, the sum type
has to specify two things:

-   The set of all possible alternatives, including their names and parameter
    types, so that the compiler can typecheck the `match` body, identify any
    unreachable `case`s, and determine whether any `case`s are missing.
-   The algorithm that, given a value of the sum type, determines which
    alternative is present, and specifies the values of its parameters.

Here's how that would look if `Optional` were defined as a class:

```carbon
class Optional(T:! Type) {
  // Factory functions
  fn Some(value: T) -> Self;
  let None:! Self;

  private var has_value: bool;
  private var value: T;

  interface MatchContinuation {

>     let template ReturnType:! Type;

Reviewer 3: What effect does the `template` keyword have here?

Author: Possibly none. This is taken from #157, but don't remember why I thought it was necessary, and it looks like @josh11b [didn't think it was](https://github.com/carbon-language/carbon-lang/pull/157/files/e4bb86dc77c0cd985e8f497233af22d26703fea9#r565653508). I'll remove it.

However, I've realized I should show the definition of `Match`, and I think `Match.Continuation` needs to be defined as a template in order to allow us to refer to `Continuation.ReturnType` in the signature of `Op` (and maybe to allow the compiler to see the method signatures it's implementing?)

    fn Some[addr me: Self*](value: T) -> ReturnType;

>     fn None[addr me: Self*]() -> ReturnType;

Reviewer 3: How do we distinguish between the case where we want to support a `.None()` pattern and the case where we want to support a `.None` pattern?

Author: I didn't have an answer to this, but I stumbled on one while responding to some of your other comments. See below.

  }

>   external impl as Match(MatchContinuation) {

Reviewer 3: I think the interface needs to be an associated constant here, not a parameter. We don't in general have a way to answer the question, "does this type implement `Match(C)` for some `C`?".

Author: Yeah, you're probably right, but that's a significant enough change that I've added the status quo to "Alternatives Considered".

    fn Op[me: Self, Continuation:! MatchContinuation](
        continuation: Continuation*) -> Continuation.ReturnType {
      if (me.has_value) {
        return continuation->Some(me.value);
      } else {
        return continuation->None();
      }
    }
  }

  // Operations like destruction, copying, assignment, and comparison are
  // omitted for brevity.
}
```

In this code, `Optional` makes itself available for use in pattern matching by
declaring that it implements the `Match` interface. `Match` takes an interface
argument, called the _continuation interface_, which specifies the set of
possible alternatives by declaring a method for each one. In this case, we pass
`Optional.MatchContinuation` as the continuation interface.

When compiling a `match` statement, the compiler checks that the type being
matched implements `Match(C)` for some continuation interface `C`. Then, it

> notionally transforms the `match` body into a class that implements `C`, with

Reviewer 3: In a previous design, you said that it's unspecified whether an `Optional(i32).Some(5)` pattern would perform a comparison (eg, evaluate the expression) or a pattern match (eg, call the continuation). Is that still the case?

Author: Yes, that's still the case. I've rewritten to make that explicit. Also, I realized we should require that both implementations are well-formed, so this unspecified choice doesn't affect whether the code compiles. That also helps make the mirroring recommendation a little closer to a requirement, and neatly solves the `.None` vs `.None()` issue.

one method for each `case`, and then passes that to `Match.Op` on the object

> being matched. For example, the `match` statement shown earlier might be

Reviewer 3: What pattern syntax does this apply to? Only `.Name` or `.Name(args)`? Or does this also apply to `Optional(T).Name` and `Optional(T).Name(args)` (presumably looking for `Optional(T) as Match` regardless of the type of the scrutinee)? Does this apply only to proper patterns or also to expression patterns?

The decision here might motivate having a specific introducer syntax for expression patterns.

Author: This applies to both syntaxes. The bare designator syntax is purely a shorthand, and is semantically equivalent to the fully-qualified syntax (see [here](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0157.md#separate-syntaxes)). I'd rather not get into that in this doc, because I suspect the bare designator syntax is going to be a can of worms.

As noted on the previous comment, it's unspecified whether this applies to expression patterns.

transformed into:

```carbon
class __MatchStatementImpl {
  fn Make() -> Self { return {}; }

  impl as Match(Optional.MatchContinuation) where .ReturnType = () {
    fn Some(the_value: i32) {
      Print(the_value);
    }
    fn None() {
      Print("None");
    }
  }
}

my_opt.(Match.Op)(__MatchStatementImpl.Make());
```

> Thus, a `match` statement works by invoking the sum type's `Match.Op` method,

Reviewer 3: Is there any guarantee as to how many times `Match.Op` might be called, or in what order? In #2188 I said effectively that we pattern match in left-to-right, top-to-bottom order, but that repeated matches on the same value are allowed (but not guaranteed) to be optimized away into a single comparison. Reiterating that here might be nice.

Author: I don't want to focus too much on the specifics of `match`, because this doc is supposed to be generically about how _any_ pattern-matching construct will hook into sum types. Instead I've rephrased to put less emphasis on `match`; does that seem OK?

which is responsible for determining which alternative the sum object
represents, and then invoking the compiler-supplied continuation that
corresponds to that alternative. In order for this scheme to work, `Match.Op` is
required to invoke exactly one method of `MatchContinuation`, and to do so

> exactly once.

Reviewer 3: I wonder if we can somehow enforce that in the type system. We already have a guarantee that if `Match.Op` returns, it must have invoked a method of the continuation, because there's no other way to construct an instance of the return type. I think what we'd be missing is a way to make the return type non-discardable.

I'm assuming that it's OK for `Match.Op` to, for example, terminate the thread or the program, or simply not terminate, instead of running exactly one continuation exactly once, but an "if it returns" would help here.

Author: > I think what we'd be missing is a way to make the return type non-discardable.

This probably isn't worth inventing a feature for, but it's something to keep in mind for the future, in case we see a way to do this with features we're already building. I think we could do that if we could define an interface that supports moving but not destruction. That's trickier than it sounds because the typechecker would need to prove that the code never actually needs the destructor, i.e. the object is moved on all possible execution paths.

> I'm assuming that it's OK for `Match.Op` to, for example, terminate the thread or the program, or simply not terminate, instead of running exactly one continuation exactly once, but an "if it returns" would help here.

OK, I've rephrased to make that clearer.

Notice that the names `Some` and `None` are defined twice, once as factory
functions/constant members of `Optional` and once as methods of
`MatchContinuation`, with the same parameter types in each case. The two
effectively act as inverses of each other: the `Optional` members are used to
create `Optional` values, and the `MatchContinuation` methods are used to report
the parameter values that would create a given `Optional`. This mirroring
between expression and pattern syntax is ultimately a design choice by the type
author; there is no language-level requirement that the alternatives correspond

> to the factory functions, but it is **strongly** recommended.

Reviewer 3: Is it worth pointing out here the consequences for not following this recommendation: specifically that the behavior of expression patterns and proper patterns diverge?

Author: I've tried to spell it out a bit more as part of my overhaul of this whole discussion. How is it now?

be extended to behave like a sum type. This is much more verbose than a `choice`
declaration, but gives the author full control over the representation and class
members.

The ability to create instances of the sum type can be straightforwardly
emulated with factory functions and static constants, and the internal storage
layout will presumably involve untagged unions or some other low-level storage
primitive which hasn't been designed yet, but the key to defining a sum type's
interface is enabling it to support pattern matching. To do that, the sum type
has to specify two things:

-   The set of all possible alternatives, including their names and parameter
    types, so that the compiler can typecheck the `match` body, identify any
    unreachable `case`s, and determine whether any `case`s are missing.
-   The algorithm that, given a value of the sum type, determines which
    alternative is present, and specifies the values of its parameters.

It does so by implementing the `Match` interface, which is defined as follows:

```carbon
interface Match {
  let template Continuation:! Type;
  fn Op[me: Self, C:! Continuation](continuation: Continuation*)

>     -> Continuation.ReturnType;

Reviewer 3: I think this should be:
```suggestion
    -> C.ReturnType;
```
... but it strikes me that using the literal identifier `ReturnType` here means that the name `ReturnType` can never be the name of an alternative, which seems unfortunate. I could imagine that coming up in practice in some not-completely-contrived situation:
```
choice PartOfFunctionDeclaration {
  Name(name: String),
  Parameters(params: Pattern),
  ReturnType(type: Expression),
}
```
... but as a matter of principle I don't like having reserved identifiers that are not keywords.

I think we can address this name collision issue. Suppose we define an interface `MatchContinuation` in the standard library:
```
interface MatchContinuation {
  let ReturnType:! Type;
}
```
then we can use this here:
```suggestion
    -> C.(MatchContinuation.ReturnType);
```

A `Match` impl would then use or `extends MatchContinuation;` (if it doesn't have a colliding alternative) or  `impl as MatchContinuation;` (if it does) instead of `let ReturnType:! Type;`:

```
interface Continuation {
  extends MatchContinuation;
  // I'm not sure if the explicit `MatchContinuation.` qualifier here is necessary.
  // Does `extends` cause name lookup in this scope to look in `MatchContinuation`?
  fn Some[addr me: Self*](value: T) -> MatchContinuation.ReturnType;
  fn None[addr me: Self*]() -> MatchContinuation.ReturnType;
}
```

Author: Oh, good catch! And yeah, your solution makes sense; done.

I'm going to optimistically assume we don't need the `MatchContinuation.` qualifier on the return type of the continuation methods.

}
```

`Continuation` must itself be an interface, and its definition specifies the set
of possible alternatives: each alternative is represented as a method of that
interface. When compiling a proper pattern (or set of patterns that includes a
proper pattern, as with the cases of a `match`) whose type is a sum type, the
compiler generates an implementation of `Continuation` and passes it to
`Match.Op`. The sum type's implementation of `Match.Op` is responsible for
determining which alternative is present and what its parameters are, and

> calling the corresponding method of `continuation` with those parameters. It is
> required to call exactly one such method exactly once before returning. The

Reviewer 3: We're jumping back and forth between the compiler's responsibilities and the `Match.Op` impl's responsibilities here, so I think it'd be useful to be explicit, even though this is the most recent possible referent:
```suggestion
calling the corresponding method of `continuation` with those parameters. The `Match.Op` implementation is
required to call exactly one such method exactly once before returning. The
```

Author: Done.

compiler populates the `Continuation` method bodies with whatever code should be

> executed when the corresponding alternatives match.

Reviewer 3: Is there any guarantee as to whether the continuation includes the `match` body or not? Eg, can an implementation choose to call `Match.Op` with a continuation that just returns an index of which match block it should use, and then switch on that after `Match.Op` returns? What guarantees exist about the lifetimes of any local variables in the `Match.Op` function as compared to the `case` block?

There are some optimizations that I'd like to be allowed to perform here (eg, combining multiple matches on different things into a single jump table lookup), but their validity would depend a lot on the rules selected.

Reviewer 3: Hm. I think that with sufficiently clever code generation, we don't need this to be an observable part of the semantics after all. If we say:

-   Matching is performed left-to-right within a pattern, and `match` patterns are considered top-down.
-   Any case block is executed within the call to the continuation in the execution of a `Match.Op` call.
-   If there would be more than one `Match.Op` call for the same scrutinee, the compiler is allowed to omit all but the first.

... then I think I can still do the optimizations that I want to do. (The trick that I have in mind is that the first call to `Match.Op` either runs the code for that alternative, or returns a value saying which alternative was chosen. Subsequent matches on the same thing that can't be merged into that same call can instead look at that returned value, and rerun the `Match.Op` call to re-enter the scope even though they know which alternative will be invoked.)

Author: > Is there any guarantee as to whether the continuation includes the `match` body or not?

No, there's no guarantee; I want to err on the side of permitting those kinds of optimizations. I've added a paragraph making that explicit. There are some potential future complications here, though, so I've added a discussion to "Alternatives considered".

I've also added a note that the continuation should be invoked as a tail call, which kind of goes hand in hand with this lack of guarantee.

> What guarantees exist about the lifetimes of any local variables in the `Match.Op` function as compared to the `case` block?

We probably don't want them to be interleaved (i.e. neither disjoint nor nested), if only for debuggability, but I'm not sure we need to specify that. Other than that, there are no guarantees about how the lifetimes relate other than the ones obviously imposed by causality (e.g. the lifetimes of `Match.Op` locals must start before the lifetimes of `case` body locals start).

> * Matching is performed left-to-right within a pattern, and `match` patterns are considered top-down.
> * Any case block is executed within the call to the continuation in the execution of a `Match.Op` call.
> * If there would be more than one `Match.Op` call for the same scrutinee, the compiler is allowed to omit all but the first.

Those constraints all make sense to me (and I think they permit the in-place mutation case described in the "alternatives considered"), but I tentatively think they belong more in the docs for pattern matching than the docs for sum types.
